tests/migrations/test_commands.py

```diff
@@ -2392,20 +2392,37 @@ def test_makemigrations_with_invalid_custom_name(self):
                 "makemigrations", "migrations", "--name", "invalid name", "--empty"
             )
 
-    def test_makemigrations_check(self):
+    def test_makemigrations_check_with_changes(self):
         """
         makemigrations --check should exit with a non-zero status when
         there are changes to an app requiring migrations.
         """
-        with self.temporary_migration_module() as tmpdir:
-            with self.assertRaises(SystemExit):
-                call_command("makemigrations", "--check", "migrations", verbosity=0)
-            self.assertFalse(os.path.exists(tmpdir))
+        out = io.StringIO()
+        with (
+            self.temporary_migration_module() as tmpdir,
+            self.assertRaises(SystemExit) as cm,
+        ):
+            call_command(
+                "makemigrations",
+                "--check",
+                "migrations",
+                stdout=out,
+            )
+        self.assertEqual(cm.exception.code, 1)
+        self.assertIn("Migrations for 'migrations':", out.getvalue())
+        self.assertFalse(os.path.exists(tmpdir))
```

Comment:
This one got me puzzled for a while: while conducting the review, I did the "usual" and I changed minimal parts of the diff to ensure a test would fail. Specifically, I changed:
```diff
--- a/django/core/management/commands/makemigrations.py
+++ b/django/core/management/commands/makemigrations.py
@@ -115,7 +115,7 @@ class Command(BaseCommand):
         self.include_header = options["include_header"]
         check_changes = options["check_changes"]
         if check_changes:
-            self.dry_run = True
+            self.dry_run = False
         self.scriptable = options["scriptable"]
         self.update = options["update"]
         # If logs and prompts are diverted to stderr, remove the ERROR style.
```
and I ran the test suite, confidently expecting that the `test_makemigrations_check_with_changes` would fail. IT DID NOT! :scream: 

After the shock, I started debugging and noticed that `self.assertFalse(os.path.exists(tmpdir))` will always hold after exiting the `self.temporary_migration_module` context manager because the `tmpdir` folder is deleted on exit (this is explained by the usage of `tempfile.TemporaryDirectory()` as context manager in the implementation of `temporary_migration_module`).

So I said "well, we *just* need to move the assert to inside the `with`  block and we are golden. Guess what? it was not enough! I also had to split the multiple `with` managers into two nested calls, I think this is because the raising of the `SystemExit` is preventing the execution of the assert. So in the end I think this is the correct test:

```diff
--- a/tests/migrations/test_commands.py
+++ b/tests/migrations/test_commands.py
@@ -2398,19 +2398,17 @@ class MakeMigrationsTests(MigrationTestBase):
         there are changes to an app requiring migrations.
         """
         out = io.StringIO()
-        with (
-            self.temporary_migration_module() as tmpdir,
-            self.assertRaises(SystemExit) as cm,
-        ):
-            call_command(
-                "makemigrations",
-                "--check",
-                "migrations",
-                stdout=out,
-            )
+        with self.temporary_migration_module() as tmpdir:
+            with self.assertRaises(SystemExit) as cm:
+                call_command(
+                    "makemigrations",
+                    "--check",
+                    "migrations",
+                    stdout=out,
+                )
+            self.assertFalse(os.path.exists(tmpdir))
         self.assertEqual(cm.exception.code, 1)
         self.assertIn("Migrations for 'migrations':", out.getvalue())
-        self.assertFalse(os.path.exists(tmpdir))
 
     def test_makemigrations_check_no_changes(self):
         """
```

Reply:
Yes, that’s absolutely correct. Thanks for checking. I’ve applied your patch and added you as coauthor.

Reply:
One adjustment was actually needed, since `MigrationWriter.basedir` always creates the directory and `__init__.py`, even in dry-run mode:

```diff
diff --git tests/migrations/test_commands.py tests/migrations/test_commands.py
index dece54f509..d9ca30b002 100644
--- tests/migrations/test_commands.py
+++ tests/migrations/test_commands.py
@@ -2406,7 +2406,7 @@ def test_makemigrations_check_with_changes(self):
                     "migrations",
                     stdout=out,
                 )
-            self.assertFalse(os.path.exists(tmpdir))
+            self.assertEqual(os.listdir(tmpdir), ["__init__.py"])
         self.assertEqual(cm.exception.code, 1)
         self.assertIn("Migrations for 'migrations':", out.getvalue())

```

Perhaps we should consider this a separate bug, since it could prevent a clean run of `makemigrations --check` in the niche scenario where the migration directory doesn't exist.

Reply:
Thanks @adamchainz! I agree that the creation of the folder + `__init__.py` file is unexpected when running with `dry-run`. Do you have time to the ticket?